### PR TITLE
zmq4: dealer returns error when closing context

### PR DESCRIPTION
**Pull request by speakyourcode**

### CODE REVIEW

File: zmq4_routerdealer_test.go [hunk 1]
```diff
@@ -255,34 +255,48 @@ func TestRouterWithNoDealer(t *testing.T) {
...
+				t.Fatalf("router could not listen: %+v", err)
+			}
+			_, port, _ := net.SplitHostPort(router.Addr().String())
+			err = dealer.Dial("tcp://*:" + port)
+			if err != nil {
+				t.Fatalf("dealer could not dial: %+v", err)
+			}
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func(sock zmq4.Socket) {
```
Comment (sbinet) [inline zmq4_routerdealer_test.go]: ITSM there's a race here: depending on how fast/slow that goroutine is scheduled, `sock.Recv` might be issued before the socket is closed.
perhaps we could add another barrier before `sock.Recv`, enabled after `socks[tt.name].Close` has been called?

File: zmq4_routerdealer_test.go [hunk 2]
```diff
@@ -255,34 +255,51 @@ func TestRouterWithNoDealer(t *testing.T) {
...
+			err = dealer.Dial("tcp://*:" + port)
+			if err != nil {
+				t.Fatalf("dealer could not dial: %+v", err)
+			}
+			start := make(chan bool)
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func(sock zmq4.Socket, start chan<- bool) {
+				defer wg.Done()
+				start <- true
```
Comment (sbinet) [inline zmq4_routerdealer_test.go]:
```suggestion
				<-start
```

File: zmq4_routerdealer_test.go [hunk 3]
```diff
@@ -255,34 +255,51 @@ func TestRouterWithNoDealer(t *testing.T) {
...
+				_, err := sock.Recv()
+				if err == nil {
+					t.Error("expected error: context canceled")
+				}
+			}(socks[tt.name], start)
 
-	err = router.Close()
-	if err != nil {
-		t.Fatalf("could not close router: %+v", err)
+			<-start
```
Comment (sbinet) [inline zmq4_routerdealer_test.go]:
```suggestion
			
```

File: zmq4_routerdealer_test.go [hunk 4]
```diff
@@ -255,34 +255,51 @@ func TestRouterWithNoDealer(t *testing.T) {
...
+			err = socks[tt.name].Close()
+			if err != nil {
+				t.Fatalf("could not close %s: %+v", tt.name, err)
+			}
+			wg.Wait()
```
Comment (sbinet) [inline zmq4_routerdealer_test.go]:
```suggestion
			start <- true
			wg.Wait()
```

File: zmq4_routerdealer_test.go (hunk 2 above)
Comment (speakyourcode) [reply zmq4_routerdealer_test.go]: No, it's the other way around. I want to check that a socket blocked on `Recv` is interrupted and returns an error when closing the connection.

Comment (speakyourcode) [reply zmq4_routerdealer_test.go]: The effect is the same, so I'll change it how you suggested

**Pull request merged**